transport/http/agent.go

```diff
@@ -0,0 +1,138 @@
+package http
+
+import (
+	"net/http"
+	"sync"
+	"time"
+
+	"go.uber.org/yarpc/transport"
+	"go.uber.org/yarpc/transport/internal/errors"
+	"go.uber.org/yarpc/transport/peer/hostport"
+)
+
+type agentConfig struct {
+	keepAlive time.Duration
+}
+
+var defaultAgentConfig = agentConfig{keepAlive: 30 * time.Second}
+
+// AgentOption customizes the behavior of an HTTP agent.
+type AgentOption func(*agentConfig)
+
+// AgentKeepAlive specifies the keep-alive period for the network connection. If
+// zero, keep-alives are disabled.
+//
+// Defaults to 30 seconds.
+func AgentKeepAlive(t time.Duration) AgentOption {
+	return func(c *agentConfig) {
+		c.keepAlive = t
+	}
+}
+
+// NewAgent creates a new http agent for managing peers and sending requests
+func NewAgent(opts ...AgentOption) *Agent {
+	cfg := defaultAgentConfig
+	for _, o := range opts {
+		o(&cfg)
+	}
+
+	return &Agent{
+		client:    buildAgentClient(&cfg),
+		peerNodes: make(map[string]*peerNode),
+	}
+}
+
+// Agent keeps track of http peers and the associated client with which the peer will call into.
+type Agent struct {
+	lock sync.RWMutex
+
+	client    *http.Client
+	peerNodes map[string]*peerNode
+}
+
+// peerNode keeps track of a HostPortPeer and any subscribers retaining it
+type peerNode struct {
+	peer        *hostport.Peer
+	subscribers map[transport.PeerSubscriber]struct{}
+}
+
+// RetainPeer gets or creates a Peer for the specificed PeerList
+func (a *Agent) RetainPeer(pid transport.PeerIdentifier, sub transport.PeerSubscriber) (transport.Peer, error) {
```

Comment:
What's PeerList in the comment?
Function name: maybe "GetOrCreate" is more explicit?